### PR TITLE
RavenDB-6669 - support for 'session.Advanced.AddOrPatch..'

### PR DESCRIPTION
**Pull request by omerlewitz**
RavenDB-6669 - support for 'session.Advanced.AddOrPatch..'

https://issues.hibernatingrhinos.com/issue/RavenDB-6669

### CODE REVIEW

File: src/Raven.Client/Documents/Session/DocumentSession.cs [hunk 1]
```diff
@@ -31,6 +32,11 @@ public partial class DocumentSession : InMemoryDocumentSessionOperations, IDocum
         /// </remarks>
         public IAdvancedSessionOperations Advanced => this;
 
+        public void AddOrPatch<T, TU>(string id, T entity, Expression<Func<T, List<TU>>> patch, Expression<Action<List<TU>>> value)
+        {
+            throw new NotImplementedException();
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/DocumentSession.cs]: ??

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs [hunk 2]
```diff
@@ -32,8 +32,14 @@ public partial interface IAdvancedSessionOperations
         void Patch<T, TKey, TValue>(T entity, Expression<Func<T, IDictionary<TKey, TValue>>> path,
             Expression<Func<JavaScriptDictionary<TKey, TValue>, object>> dictionaryAdder);
 
-        void Patch<T, TKey, TValue>(string id, Expression<Func<T, IDictionary<TKey, TValue>>> path,
+        void Patch<T, TKey, TValue>(string id, Expression<Func<T, IDictionary<TKey, TValue>>> patch,
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]:
```suggestion
        void Patch<T, TKey, TValue>(string id, Expression<Func<T, IDictionary<TKey, TValue>>> path,
```

This is the path to patch, not a patch itself

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 3]
```diff
@@ -40,11 +41,7 @@ public abstract partial class InMemoryDocumentSessionOperations
             var variable = $"this.{pathScript}";
             var value = $"args.val_{_valsCount}";
 
-            var patchRequest = new PatchRequest
-            {
-                Script = $"{variable} = {variable} ? {variable} + {value} : {value};",
-                Values = { [$"val_{_valsCount}"] = valToAdd }
-            };
+            var patchRequest = new PatchRequest {Script = $"{variable} = {variable} ? {variable} + {value} : {value};", Values = {[$"val_{_valsCount}"] = valToAdd}};
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: Format this properly

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 4]
```diff
@@ -54,6 +51,94 @@ public abstract partial class InMemoryDocumentSessionOperations
...
 
+        public void AddOrIncrement<T, TU>(string id, T entity, Expression<Func<T, TU>> patch, TU valToAdd)
+        {
+            
+            var pathScript = patch.CompileToJavascript(_javascriptCompilationOptions);
+
+            var variable = $"this.{pathScript}";
+            var value = $"args.val_{_valsCount}";
+
+            var patchRequest = new PatchRequest {Script = $"{variable} = {variable} ? {variable} + {value} : {value};", Values = {[$"val_{_valsCount}"] = valToAdd}};
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: Format properly

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 5]
```diff
@@ -67,11 +152,7 @@ public abstract partial class InMemoryDocumentSessionOperations
 
             var valueToUse = AddTypeNameToValueIfNeeded(path.Body.Type, value);
 
-            var patchRequest = new PatchRequest
-            {
-                Script = $"this.{pathScript} = args.val_{_valsCount};",
-                Values = { [$"val_{_valsCount}"] = valueToUse }
-            };
+            var patchRequest = new PatchRequest {Script = $"this.{pathScript} = args.val_{_valsCount};", Values = {[$"val_{_valsCount}"] = valueToUse}};
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs:187]: Formatting

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 6]
```diff
@@ -192,19 +267,12 @@ private bool TryMergePatches(string id, PatchRequest patchRequest)
                 newVals[kvp.Key] = kvp.Value;
             }
 
-            Defer(new PatchCommandData(id, null, new PatchRequest
-            {
-                Script = newScript,
-                Values = newVals
-            }, null));
+            Defer(new PatchCommandData(id, null, new PatchRequest {Script = newScript, Values = newVals}, null));
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs:303]: In general, you need to fix the IDE settings on those to match our coding conventions

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Session/DocumentSession.cs (hunk 1 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/DocumentSession.cs]: done

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs (hunk 2 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 3 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 4 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 5 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs:187]: done

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 6 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs:303]: 👌

**Review by ayende — APPROVED** [review]

File: src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs [hunk 7]
```diff
@@ -9,16 +10,21 @@ namespace Raven.Client.Documents.Commands.Batches
 {
     public class PatchCommandData : ICommandData
     {
-        public PatchCommandData(string id, string changeVector, PatchRequest patch, PatchRequest patchIfMissing)
+        public PatchCommandData()
+        {
+            
+        }
+
+        public PatchCommandData(string id, string changeVector, PatchRequest patch, PatchRequest patchIfMissing = null)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs]: This changes the signature of the ctor and is considered a breaking change

File: src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs [hunk 8]
```diff
@@ -9,16 +10,21 @@ namespace Raven.Client.Documents.Commands.Batches
 {
     public class PatchCommandData : ICommandData
     {
-        public PatchCommandData(string id, string changeVector, PatchRequest patch, PatchRequest patchIfMissing)
+        public PatchCommandData()
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs]: I do not like it, create a dedicated ctor

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs [hunk 9]
```diff
@@ -31,9 +31,12 @@ public partial interface IAdvancedSessionOperations
 
         void Patch<T, TKey, TValue>(T entity, Expression<Func<T, IDictionary<TKey, TValue>>> path,
             Expression<Func<JavaScriptDictionary<TKey, TValue>, object>> dictionaryAdder);
-
-        void Patch<T, TKey, TValue>(string id, Expression<Func<T, IDictionary<TKey, TValue>>> path,
-            Expression<Func<JavaScriptDictionary<TKey, TValue>, object>> dictionaryAdder);
+        
+        void AddOrPatch<T, TU>(string id, T entity, Expression<Func<T, TU>> patch, TU value);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]: patch => path

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs [hunk 10]
```diff
@@ -31,9 +31,12 @@ public partial interface IAdvancedSessionOperations
...
+        
+        void AddOrPatch<T, TU>(string id, T entity, Expression<Func<T, List<TU>>> patch, Expression<Func<JavaScriptArray<TU>,object>> arrayAdder);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]: patch => path

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs [hunk 11]
```diff
@@ -31,9 +31,12 @@ public partial interface IAdvancedSessionOperations
...
+        
+        void AddOrIncrement<T, TU>(string id, T entity, Expression<Func<T, TU>> patch, TU valToAdd);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]: patch => path

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs [hunk 12]
```diff
@@ -31,9 +31,12 @@ public partial interface IAdvancedSessionOperations
 
         void Patch<T, TKey, TValue>(T entity, Expression<Func<T, IDictionary<TKey, TValue>>> path,
             Expression<Func<JavaScriptDictionary<TKey, TValue>, object>> dictionaryAdder);
-
-        void Patch<T, TKey, TValue>(string id, Expression<Func<T, IDictionary<TKey, TValue>>> path,
-            Expression<Func<JavaScriptDictionary<TKey, TValue>, object>> dictionaryAdder);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs:36]: Why this was deleted?

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 13]
```diff
@@ -39,21 +40,137 @@ public abstract partial class InMemoryDocumentSessionOperations
...
+                        [Constants.Documents.Metadata.RavenClrType] = clrType
+                    }
+                });
+            
+            _valsCount++;
+
+            Defer(new PatchCommandData(id,
+                null,
+                patchRequest)
+                  {
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: formatting?

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 14]
```diff
@@ -39,21 +40,137 @@ public abstract partial class InMemoryDocumentSessionOperations
...
+                        [Constants.Documents.Metadata.Collection] = collectionName, [Constants.Documents.Metadata.RavenClrType] = clrType
+                    }
+                });
+            
+            _valsCount++;
+
+            Defer(new PatchCommandData(id, null, patchRequest) {CreateIfMissing = newInstance});
+        }
+
+        public void AddOrPatch<T, TU>(string id, T entity, Expression<Func<T, TU>> patch, TU value)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: patch => path

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 15]
```diff
@@ -39,21 +40,137 @@ public abstract partial class InMemoryDocumentSessionOperations
...
+
+            Defer(new PatchCommandData(id,
+                null,
+                patchRequest)
+                  {
+                CreateIfMissing = newInstance
+            });
+        }
+
+        public void AddOrPatch<T, TU>(string id, T entity, Expression<Func<T, List<TU>>> patch, Expression<Func<JavaScriptArray<TU>, object>> arrayAdder)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: patch => path

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs [hunk 16]
```diff
@@ -39,21 +40,137 @@ public abstract partial class InMemoryDocumentSessionOperations
...
             {
-                Defer(new PatchCommandData(id, null, patchRequest, null));
+                Defer(new PatchCommandData(id,
+                    null,
+                    patchRequest,
+                    null));
             }
         }
 
+        public void AddOrIncrement<T, TU>(string id, T entity, Expression<Func<T, TU>> patch, TU valToAdd)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: patch => path

Comment (ppekrol) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: valToAdd => valueToAdd

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs (hunk 9 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs (hunk 10 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs (hunk 11 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs (hunk 12 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IAdvancedSessionOperations.Patch.cs:36]: I will take a closer look at my deletions next time

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 13 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 14 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 15 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs (hunk 16 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs]: done

File: src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs (hunk 7 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs]: done

File: src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs (hunk 8 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Commands/Batches/PatchCommandData.cs]: done

**Review by ppekrol — APPROVED** [review]

**Review by ayende — APPROVED** [review]

**Pull request merged**